test/core/event_engine/test_suite/timer_test.cc

```diff
@@ -0,0 +1,158 @@
+    auto handle = engine->RunAt(absl::InfiniteFuture(), [this]() {
+      grpc_core::MutexLock lock(&mu_);
+      signaled_ = true;
+    });
+    ASSERT_TRUE(engine->Cancel(handle));
+  }
+  // The engine is deleted, and all closures should have been flushed
+  grpc_core::MutexLock lock(&mu_);
+  ASSERT_FALSE(signaled_);
+}
+
+TEST_F(EventEngineTimerTest, TimersRespectScheduleOrdering) {
+  // Note: this is a brittle test if the first call to `RunAt` takes longer than
+  // the second callback's wait time.
+  std::vector<uint8_t> ordered;
+  uint8_t count = 0;
+  grpc_core::MutexLock lock(&mu_);
+  {
+    auto engine = this->NewEventEngine();
+    engine->RunAt(absl::Now() + absl::Seconds(1), [&]() {
+      grpc_core::MutexLock lock(&mu_);
+      ordered.push_back(2);
+      ++count;
+      cv_.Signal();
+    });
+    engine->RunAt(absl::Now(), [&]() {
+      grpc_core::MutexLock lock(&mu_);
+      ordered.push_back(1);
+      ++count;
+      cv_.Signal();
+    });
+    // Ensure both callbacks have run. Simpler than a mutex.
+    while (count != 2) {
+      cv_.WaitWithTimeout(&mu_, absl::Microseconds(100));
+    }
+  }
+  // The engine is deleted, and all closures should have been flushed beforehand
+  ASSERT_THAT(ordered, ElementsAre(1, 2));
+}
+
+TEST_F(EventEngineTimerTest, CancellingExecutedCallbackIsNoopAndReturnsFalse) {
+  auto engine = this->NewEventEngine();
+  grpc_core::MutexLock lock(&mu_);
+  auto handle = engine->RunAt(absl::Now(), [this]() {
+    grpc_core::MutexLock lock(&mu_);
+    signaled_ = true;
+    cv_.Signal();
+  });
+  cv_.WaitWithTimeout(&mu_, absl::Seconds(10));
+  ASSERT_TRUE(signaled_);
+  // The callback has run, and now we'll try to cancel it.
+  ASSERT_FALSE(engine->Cancel(handle));
+}
+
+void EventEngineTimerTest::ScheduleCheckCB(absl::Time when,
+                                           std::atomic<int>* call_count,
+                                           std::atomic<int>* fail_count,
+                                           int total_expected) {
+  // TODO(hork): make the EventEngine the time source of truth! libuv supports
+  // millis, absl::Time reports in nanos. This generic test will be hard-coded
+  // to the lowest common denominator until EventEngines can compare relative
+  // times with supported resolution.
+  int64_t now_millis = absl::ToUnixMillis(absl::Now());
+  int64_t when_millis = absl::ToUnixMillis(when);
+  EXPECT_LE(when_millis, now_millis);
+  if (when_millis > now_millis) ++(*fail_count);
+  if (++(*call_count) == total_expected) {
+    grpc_core::MutexLock lock(&mu_);
+    signaled_ = true;
+    cv_.Signal();
+  }
+}
+
+TEST_F(EventEngineTimerTest, StressTestTimersNotCalledBeforeScheduled) {
+  auto engine = this->NewEventEngine();
+  int thread_count = 100;
+  float timeout_min_seconds = 1;
+  float timeout_max_seconds = 10;
+  std::atomic<int> call_count{0};
+  std::atomic<int> failed_timer_count{0};
+  absl::BitGen bitgen;
+  for (int i = 0; i < thread_count; ++i) {
```

Comment:
s/timer_count/thread_count through here...

We probably want 100 threads creating 100 timers, for 10000 timers scheduled total. (i.e. 100 threads each run this loop).

This will help catch any threading issues with scheduling new timers too.

Reply:
I've added the 10k `RunAt` calls (split across 100 threads). As a result, I decided to delete the SimpleEventEngine example implementation rather than improve it, because TSAN was not able to deal with 10k threads in that naive implementation. The LibuvEventEngine PR will exercise this suite, and at the very least, the SimpleEventEngine implementation was helpful to verify the suite up to this point.